### PR TITLE
Jhbate/timings

### PR DESCRIPTION
**Pull request by jh-bate**
ping @darinkrauss 

I will do some initial testing on this in development and then if that looks good then will ask for review prior to merge. 

### Updates:

 - addition of logging for time taken for service call and also mongo query. 
 - simplify how we do the mongo query to exclude fields we don't want to return

### CODE REVIEW

File: tide-whisperer.go [hunk 1]
```diff
@@ -114,33 +117,37 @@ func main() {
 	}
 	_ = session.DB("").C(deviceDataCollection).EnsureIndex(index)
 
-	defer session.Close()
```
Comment (darinkrauss) [inline tide-whisperer.go]: Do we still want to defer close the session?

File: tide-whisperer.go [hunk 2]
```diff
@@ -114,33 +117,37 @@ func main() {
...
-
 	router := pat.New()
-	router.Add("GET", "/status", http.HandlerFunc(func(res http.ResponseWriter, _ *http.Request) {
+	router.Add("GET", "/status", http.HandlerFunc(func(res http.ResponseWriter, req *http.Request) {
```
Comment (darinkrauss) [inline tide-whisperer.go]: Still don't need req argument here.

File: tide-whisperer.go [hunk 3]
```diff
@@ -149,45 +156,52 @@ func main() {
...
-			log.Print("Iterator ended with an error: ", err)
-			failureReturnCode = 500
+
+		log.Println(DATA_API_PREFIX, fmt.Sprintf("mongo query took [%.5f]secs and returned [%d] records", time.Now().Sub(startQueryTime).Seconds(), len(results)))
+
+		if len(results) == 0 {
+			log.Println(DATA_API_PREFIX, fmt.Sprintf("completed in [%.5f]secs", time.Now().Sub(start).Seconds()))
+			log.Println(DATA_API_PREFIX, query_no_data)
+			res.Header().Add("content-type", "application/json")
+			res.WriteHeader(http.StatusNotFound)
```
Comment (darinkrauss) [inline tide-whisperer.go]: Should this use the http.Error(res, ...) mechanism or perhaps http.NotFound(...)?

File: tide-whisperer.go [hunk 4]
```diff
@@ -149,45 +156,52 @@ func main() {
...
+
+		log.Println(DATA_API_PREFIX, fmt.Sprintf("mongo query [%#v]", groupDataQuery))
+
+		startQueryTime := time.Now()
+
+		//return un-ordered (i.e. the order isn't guaranteed by mongo)
+		err := mongoSession.DB("").C(deviceDataCollection).
+			Find(groupDataQuery).
+			Select(removeFieldsForReturn).
+			All(&results)
```
Comment (darinkrauss) [inline tide-whisperer.go]: We already talked about the potential issue with All with large data sets.

File: tide-whisperer.go [hunk 5]
```diff
@@ -114,33 +117,37 @@ func main() {
...
+	router.Add("GET", "/status", http.HandlerFunc(func(res http.ResponseWriter, req *http.Request) {
 		if err := session.Ping(); err != nil {
-			res.WriteHeader(500)
-			res.Write([]byte(err.Error()))
+			log.Println(DATA_API_PREFIX, "status", err.Error())
+			http.Error(res, err.Error(), http.StatusInternalServerError)
 			return
 		}
-		res.WriteHeader(200)
+		res.WriteHeader(http.StatusOK)
```
Comment (darinkrauss) [inline tide-whisperer.go]: Isn't the default 200 (so we don't even need this header, I think)?

**Comment by darinkrauss** [comment]
A few comments, but overall looks good.  I think the Mongo query is nicely cleaned up.  

Once we get rid of groupId, we'll want to update the index to add the _active field and double check that Mongo is not scanning more records than it needs to.

File: tide-whisperer.go (hunk 5 above)
Comment (jh-bate) [inline tide-whisperer.go]: yes - you are correct

File: tide-whisperer.go (hunk 1 above)
Comment (jh-bate) [inline tide-whisperer.go]: maybe - I will check this

File: tide-whisperer.go (hunk 3 above)
Comment (jh-bate) [inline tide-whisperer.go]: yeah to be consistent probably should be  200 and an empty array - I will just need to consider if there is an impact on current usage in blip / platform client

File: tide-whisperer.go [hunk 6]
```diff
@@ -99,6 +117,23 @@ func main() {
 		return !(perms["root"] == nil && perms["view"] == nil)
 	}
 
+	//log error detail and write as application/json
+	jsonError := func(res http.ResponseWriter, err detailedError, startedAt time.Time) {
+
+		if err.RawMessage == "" {
```
Comment (darinkrauss) [inline tide-whisperer.go]: You can probably get rid of this and just use "omitempty" on the RawMessage json serialization.  No need to add duplicate data to the logs.

Comment (jh-bate) [inline tide-whisperer.go]: so - don't bother with the copy or even having the RawMessage filed and just pass the original error if one exists and log that 

e.g. 
```
jsonError := func(res http.ResponseWriter, err detailedError, internal error, startedAt time.Time) {
    ...

    log.Println(DATA_API_PREFIX, fmt.Sprintf("[%s] [%s] failed after [%.5f]secs with error [%s] ", err.Id, err.Code, time.Now().Sub(startedAt).Seconds(), internal.Error()))

    ...

}
```

Comment (darinkrauss) [inline tide-whisperer.go]: Oh, you could do that.  I have no problem with keeping the RawMessage as part of the error object.  Might want to call it "InternalData"/"InternalMessage"/"LogMessage" to indicate it is not sent back.  I was just saying that there wasn't any point of copying the Message to RawMessage if you also logged Message.

So,

`log.Println(DATA_API_PREFIX, fmt.Sprintf("[%s] [%s] failed after [%.5f]secs with error [%s][%s] ", err.Id, err.Code, time.Now().Sub(startedAt).Seconds(), err.Message, err.RawMessage))`

Comment (jh-bate) [inline tide-whisperer.go]: got it - thanks

**Pull request merged**